conda_forge_tick/make_graph.py

```diff
@@ -48,10 +48,19 @@ def get_attrs(name, i, bad):
         req = set([x.split()[0] for x in req])
     sub_graph['req'] = req
 
-    keys = [('source', 'url'), ('package', 'name'),
-            ('package', 'version')]
+    keys = [('package', 'name'), ('package', 'version')]
     missing_keys = [k[1] for k in keys if k[1] not in
                     yaml_dict.get(k[0], {})]
+    source = yaml_dict.get('source', [])
+    if not isinstance(source, list):
```

Comment:
It would be preferable if this was `not isinstance(source, collections.abc.Sequence)` or `isinstance(source, str)`.  What are the types that source could come in as?

Reply:
yes, isinstance is fragile.  In conda-build, I have a function called "ensure_list" that is a no-op for lists and tuples, but does this same enclosure of a string on lists.

https://github.com/conda/conda-build/blob/master/conda_build/utils.py#L978-L984

Not sure exactly how much better that is, but the hasattr might be less fragile than a purely isinstance-based approach.

Reply:
Here source will either be a dict or a list, not a string. I don't think `ensure_list` will work in this case.

Reply:
In light of that, I think you want

```
if hasattr(source, "keys"):
```

Reply:
actually, I think in that case you want to check if it is a mapping:

`if isinstance(source, collections.abc.Mapping):`

Reply:
Yes, that's a stronger check.